File: leet_code/unique_paths.py
class Solution:
    moves = [(-1, 0), (1, 0), (0, 1), (0, -1)]
    target_vals = {0, 2}
    
    def uniquePathsIII(self, grid) -> int:
        self.grid = grid
        self.start = None
        self.unseen = set()
        self.memo = dict()
        self.find_unseen_nodes()
        return self.uniquePathsIIIHelper(self.start[0], self.start[1])
        
    
    def find_unseen_nodes(self):
        for i in range(len(self.grid)):
            for j in range(len(self.grid[i])):
                if self.grid[i][j] in Solution.target_vals:
                    self.unseen.add((i, j))
                elif self.grid[i][j] == 1:
                    self.start = (i, j)
        assert(self.start is not None)
        
        
    def uniquePathsIIIHelper(self, start_row, start_col):
        total = 0
        for drow, dcol in Solution.moves:
            new_row, new_col = start_row + drow, start_col + dcol
            if not self.in_bound(new_row, new_col): continue
            
            # Case 1: visit unvisited node, make move
            if self.grid[new_row][new_col] == 0:
                self.grid[new_row][new_col] = 1
                self.unseen.remove((new_row, new_col))

                result = self.uniquePathsIIIHelper(new_row, new_col)
                total += result
                
                self.grid[new_row][new_col] = 0
                self.unseen.add((new_row, new_col))
            
            # Case 2: visit goal
            elif self.grid[new_row][new_col] == 2:
                if len(self.unseen) == 1 and (new_row, new_col) in self.unseen:
                    total += 1
                    break

        return total
                
    def in_bound(self, row, col):
        return ((0 <= row and row < len(self.grid)) and 
                (0 <= col and col < len(self.grid[row])))

A = Solution()
print(A.uniquePathsIII([[1, 0, 0, 0],
                  [0, 0, 0, 0],
                  [0, 0, 2, -1]]))

File: leet_code/unique_paths.cpp
#include <bits/stdc++.h> 
#include <algorithm>
typedef vector<vector<int>> vec_tuples;

// NOTE: doesn't compile, "double-free error"
class Solution {
    vec_tuples moves {
        {1, 0}, {0, -1}, {0, 1}, {0, 1}
    };
    vec_tuples grid;
    vec_tuples unseen;
    
    int in_bound(int row, int col) {
        return ((0 <= row && row < this->grid.size()) && 
                (0 <= col && col < this->grid[row].size()));
    }
    
    int paths_helper(int row, int col) {
        int total = 0;
        for (vector<int> move : moves) {
            int new_row = row + move[0];
            int new_col = col + move[1];
            if (!in_bound(new_row, new_col)) continue;
            vector<int> new_pos{new_row, new_col};
            
            // Case 1: visit unvisited node, make move recursive
            if (this->grid[new_row][new_col] == 0) {
                // make move
                this->grid[new_row][new_col] = 1;
                auto it = std::find(this->unseen.begin(), this->unseen.end(), 
                                    new_pos);
                std::swap(*it, this->unseen.back());
                this->unseen.pop_back();  // remove 
                
                total += this->paths_helper(new_row, new_col);
                
                // undo move
                this->grid[new_row][new_col] = 0;
                this->unseen.push_back(new_pos);
            }
            
            // Case 2: visit goal
            else if (this->grid[new_row][new_col] == 2) {
                // if seen all nodes, done, return 1
                if (this->unseen.size() == 1 && this->unseen[0] == new_pos) {
                    total++;
                }
            }
        }
        return total;
    }
public:
    int uniquePathsIII(vec_tuples& grid) {
        // generate unseen vector
        this->grid = grid;
        
        int start_r = 0, start_c = 0;
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid.size(); j++) {
                if (grid[i][j] == 0 || grid[i][j] == 2) {
                    vector<int> pos{i, j};
                    this->unseen.push_back(pos);
                }
                else if (grid[i][j] == 1) {
                    start_r = i;
                    start_c = j;
                }
            }
        }
        return this->paths_helper(start_r, start_c);
    }
};
